File: js/data_structures/graph.js
/**
 * 图结构
 * 单向图，多向图
 * 图由顶点与边组成,
 * 图的信息都在边上
 */

const Queue = require('./queue')

// 图的顶点
// class Vertex {
//     constructor(label, isVisited = false) {
//         this.label = label
//         this.isVisited = isVisited
//     }
// }

// 图
class Graph {
    /**
     * 采用邻接列表方式或者邻接矩阵方式实现
     * 每一个顶点会存储一个从它这里开始的边的列表
     * 1 -> [2]
     * 2 -> [1, 3, 4]
     * 3 -> [2]
     * 4 -> [2]
     */
    constructor(verticles, vertexList = []) {
        // ?
        this.verticles = verticles
        this.edges = 0
        this.adj = []
        this.visited = []; // 存储已遍历的顶点数组
        this.edgeTo = []; // 从一个顶点到另一个顶点到边
        this.searchedList = []; // 搜索遍历后的数组
        this.vertexList = vertexList; // 需要用来进行排序的集合
        this.initAdjList()
    }
    // 初始化邻接列表 Adjacency List
    initAdjList() {
        /**
         * 初始化一个多维数组列表
         */
        for (let i = 0; i < this.verticles; i++) {
            this.adj[i] = [];
            this.visited[i] = false;
        }
    }
    // 添加一条边(两点决定一条边)
    addEdge(a, b) {
        // a顶点添加一个对b的索引
        this.adj[a].push(b);
        // b顶点添加一个对a的索引
        this.adj[b].push(a);
        // 边数加1
        this.edges++;
    }
    /**
     * deep first search
     * 深度优先搜索, 每遍历一个元素，都尽可能地把这个元素的索引都遍历完
     */
    dfs(v) {

        this.visited[v] = true;
        if (this.adj[v]) {
            this.searchedList.push(v)
        }

        for (let i = 0; i < this.adj[v].length; i++) {
            const curItem = this.adj[v][i];
            if (!this.visited[curItem]) { // 该顶点没有被访问过
                this.dfs(curItem)
            }
        }

    }
    /**
     * breadth first search
     * 广度优先搜索
     */
    bfs(v) {
        // 能找到这个顶点
        if (this.adj[v]) {
            /**
             * 使用队列来实现，将已经访问过的顶点入队
             * 最后全部出队即可
            */
            const vertexQueue = new Queue()
            this.visited[v] = true;
            vertexQueue.put(v);
            while (vertexQueue.size()) {
                // 队首元素出队
                const vert = vertexQueue.get();
                this.searchedList.push(vert);
                // 开始访问这个顶点中所关联的其他顶点
                for (const item of this.adj[vert]) {
                    if (!this.visited[item]) {
                        this.visited[item] = true;
                        // 正在探索到顶点，添加一条边
                        this.edgeTo[item] = vert;
                        vertexQueue.put(item)
                    }
                }
            }

        }
    }
    // 从一个顶点到达某一个顶点的最短路径算法
    pathTo(from, to) {

        // 先对该顶点进行一次搜索遍历, 记录下所有对路径点
        this.resetSearchList();
        this.bfs(from);

        const path = [];
        // 是否有路径经过这个点
        if (!this.visited[to]) {
            return null;
        }
        for (let i = to; i != from; i = this.edgeTo[i]) {
            path.unshift(i)
        }
        path.unshift(from)
        return path
    }
    /**
     * 拓扑排序, 有先决条件
     * 针对有向无环图实现拓扑排序
     * 其目的是为了处理图里大小分支的优先级与顺序，
     * 是一种非稳定性的排序，二级排序的顺序可以不是唯一的
     * 从第n个顶点开始排序
     */
    topSort(n = 0) {
        const stack = [];
        // 重置访问节点信息
        this.resetSearchList();
        this.dfs(n);
        this.searchedList.forEach(item => {
            stack.push(this.vertexList[item])
        })
        return stack
    }
    // 清空搜索列表
    resetSearchList() {
        this.searchedList = [];
        // 重置访问记录
        for (let i = 0; i < this.verticles; i++) {
            this.visited[i] = false;
        }
    }
    // 展示图
    showMap(i) {
        if (i >= 0) {
            return this.adj[i]
        }
        return this.adj
    }
}

module.exports = Graph

File: py/data_structures/graph.py
# 图与图算法
from queue import Queue


# 图类, 使用邻接列表或者邻接矩阵存储
# 使用邻接列表方式存储已经在js中实现，py中尝试使用邻接矩阵
class Graph:
    def __init__(self, vertex_list, is_direc=False):
        # 是否是有向图
        self.is_direc = is_direc
        self.verticles = len(vertex_list) or 0
        # 邻接矩阵存储
        self.metrix = []
        # 顶点存储
        self.vertex = {}
        # 边存储, 创建一个len x len的矩阵, 存储最大图到边
        self.edge_to = [None for item in range(self.verticles)]
        # 搜索后的排序数组
        self.search_list = []
        # top排序的stack
        self.top_stack = []
        self.init_metrix(vertex_list)

    # 初始化邻接矩阵
    def init_metrix(self, vertex_list):
        for i in range(0, self.verticles):
            # 顺便将顶点初始化
            if vertex_list and vertex_list[i]:
                vertex = Vertex(vet_id=i, data=vertex_list[i])
                self.vertex["vertex%d" % i] = vertex
            self.metrix.append([])
            for j in range(0, self.verticles):
                self.metrix[i].append(0)

    # 给两个顶点添加一条边(区分无向图和有向图)
    def add_edge(self, from_id, to_id):
        if from_id < self.verticles and to_id < self.verticles:
            self.metrix[from_id][to_id] = 1
            # self.vertex["vertex%d" % from_id].to_ids.append(to_id)
            # 如果是无向图, 可以同时给下面这段赋值
            if not self.is_direc:
                self.metrix[to_id][from_id] = 1
            # self.vertex["vertex%d" % to_id].from_ids.append(from_id)

    # 重置搜索遍历点
    def reset_search_list(self):
        # 清空搜索数组
        self.search_list = []
        # 边存储, 创建一个len x len的矩阵, 存储最大图到边
        self.edge_to = [None for item in range(self.verticles)]
        # 重置is_visited
        for i in range(self.verticles):
            self.vertex["vertex%d" % i].is_visited = False

    # 深度优先搜索
    def dfs(self, ver_id):

        cur_vertex = self.vertex["vertex%d" % ver_id]
        cur_vertex.is_visited = True

        # 横向遍历, for循环需要拿到索引的话, 需要用enumerate
        for to_id, item in enumerate(self.metrix[ver_id]):
            next_vertex = self.vertex["vertex%d" % to_id]
            # 有连接点, 走到下一个分支去递归遍历
            if item == 1 and not next_vertex.is_visited:
                self.dfs(to_id)
        self.search_list.append(ver_id)

    # 广度优先搜索
    def bfs(self, ver_id):

        cur_vertex = self.vertex["vertex%d" % ver_id]
        if not cur_vertex.is_visited:

            cur_vertex.is_visited = True
            # 将元素入队
            search_queue = Queue(maxsize=self.verticles)
            search_queue.put(ver_id)

            while search_queue.qsize():
                cur_id = search_queue.get()
                self.search_list.append(cur_id)
                for to_id, item in enumerate(self.metrix[cur_id]):
                    next_vertex = self.vertex["vertex%d" % to_id]
                    if item == 1 and not next_vertex.is_visited:
                        next_vertex.is_visited = True
                        self.edge_to[to_id] = cur_id
                        search_queue.put(to_id)

    # 最短路径
    def short_path(self, from_id, to_id):
        stack = []
        self.reset_search_list()
        # 先使用广度优先搜索，从该顶点走完全图
        self.bfs(from_id)
        # 遍历之后，先看看是否能走到to_id的点
        to_vertex = self.vertex["vertex%d" % to_id]
        if to_vertex and to_vertex.is_visited:
            idx = to_id
            while idx != from_id:
                edge_id = self.edge_to[idx]
                stack.append(idx)
                idx = edge_id
            stack.append(from_id)
            stack.reverse()
            return stack
        # 没有找到对应点，或者没有经过目标点，说明无法到达
        else:
            return None

    # 拓扑排序, 仅针对有向无环图
    def top_sort(self):
        top_vertex = self._find_top()
        self.top_stack = []
        # 从顶点开始深度优先搜索
        if top_vertex is not None:
            self._loop_top(top_vertex.id)
        return self.top_stack

    # 对顶点元素进行遍历
    def _loop_top(self, ver_id):
        vertex = self.vertex["vertex%d" % ver_id]
        if not vertex.is_visited:
            vertex.is_visited = True
            self.top_stack.append(vertex.data)

        for idx, item in enumerate(self.metrix[ver_id]):
            cur_vertex = self.vertex["vertex%d" % idx]
            if item == 1 and not cur_vertex.is_visited:
                self._loop_top(cur_vertex.id)

    # 查找没有前驱节点的顶点
    # 对邻接矩阵而言，即没有与矩阵中的任何顶点建立由其他顶点指向该顶点的单向的边
    # 即任何一行ℹ中都不存在row[i][j] = 1的情况, 那j就是顶点
    # 或者任何一列中, 也是可以的
    def _find_top(self):
        row_id = -1
        for row in range(self.verticles):
            col_arr = []
            for col in range(self.verticles):
                col_arr.append(self.metrix[col][row])
            col_set = set(col_arr)
            # 是一个全0的集合
            if 1 not in col_set:
                return self.vertex["vertex%d" % row]
        return None

    def show_search_list(self):
        arr = []
        while len(self.search_list):
            arr.append(self.search_list.pop())
        return arr


# 顶点类
class Vertex:
    def __init__(self, data='', vet_id=0):
        self.is_visited = False
        self.data = data
        self.id = vet_id
